js/packages/teams-ai/src/authentication/MessagingExtensionAuthentication.ts

```diff
@@ -0,0 +1,111 @@
+import { ActivityTypes, InvokeResponse, TokenResponse, TurnContext } from 'botbuilder';
+import { DefaultTurnState } from '../DefaultTurnStateManager';
+import { TurnState } from '../TurnState';
+import { FETCH_TASK_INVOKE_NAME, QUERY_INVOKE_NAME, QUERY_LINK_INVOKE_NAME } from '../MessageExtensions';
+import * as UserTokenAccess from './UserTokenAccess';
+import { OAuthPromptSettings } from 'botbuilder-dialogs';
+
+/**
+ * @internal
+ */
+export class MessagingExtensionAuthentication<TState extends TurnState = DefaultTurnState> {
+    private readonly _oauthPromptSettings: OAuthPromptSettings;
+
+    public constructor(_oauthPromptSettings: OAuthPromptSettings) {
+        this._oauthPromptSettings = _oauthPromptSettings;
+    }
+
+    public async authenticate(context: TurnContext, state: TState): Promise<string | undefined> {
+        const authObj = context.activity.value.authentication;
+
+        // Token Exchange
+        if (authObj && authObj.token) {
+            // Message extension token exchange invoke activity
+            const isTokenExchangable = await this.isTokenExchangeable(context);
+            if (!isTokenExchangable) {
+                await context.sendActivity({
+                    value: { status: 412 } as InvokeResponse,
+                    type: ActivityTypes.InvokeResponse
+                });
+
+                return undefined;
+            }
+        }
+
+        const value = context.activity.value;
+
+        // When the Bot Service Auth flow completes, the query.State will contain a magic code used for verification.
+        const magicCode = value.state && Number.isInteger(Number(value.state)) ? value.state : '';
+
+        const tokenResponse = await UserTokenAccess.getUserToken(context, this._oauthPromptSettings, magicCode);
+
+        if (!tokenResponse || !tokenResponse.token) {
+            // There is no token, so the user has not signed in yet.
+            // Retrieve the OAuth Sign in Link to use in the MessagingExtensionResult Suggested Actions
+
+            const signInResource = await UserTokenAccess.getSignInResource(context, this._oauthPromptSettings);
+            const signInLink = signInResource.signInLink;
+            // Do 'silentAuth' if this is a composeExtension/query request otherwise do normal `auth` flow.
+            const authType = context.activity.name === QUERY_INVOKE_NAME ? 'silentAuth' : 'auth';
```

Comment:
If developer does not enable SSO, using `silentAuth` (which means using SSO) may fail?

Reply:
Maybe we need a way to determine which flow (SSO / non-SSO) to use?

Reply:
Great catch. I have a feeling that if SSO isn't enabled Teams client should be able to just default to auth even if 'silentAuth' property is used.

Reply:
if we send `silentAuth` then Teams client trigger token exchange flow. When the bot tries to do token exchange it will fail and elegantly return 412 response. To which Teams client will trigger the auth flow.